React/forms.md

```diff
@@ -2,24 +2,280 @@
 
 ## Learning Goals
 - Implement a form in a component
-- Handle the event of a form submission in a component
 - Dynamically provide user feedback as they complete a form
+- Handle the event of a form submission in a component
 
 ### Overview
 Since we've already learn how event handling works in React, we should be able to take the application of this knowledge one step further to allow us to submit a set of form data. Forms work a bit different than other HTML elements because they maintain some of their own state.
 
 Additionally, one of the most powerful aspects of using any JavaScript framework is to provide some dynamic user interaction. We'll see here how to integrate ongoing user feedback, so the user does not have to wait until the form is submitted to know that they might have issues with the data.
 
+## A Regular HTML Form
+
+A normal HTML form to add a student to a list might look like this:
+
+```html
+<form id="new-student" >
+  <label for="name">Student Name</label>
+  <input name="name" type="text">
+  <label for="email">Student Email</label>
+  <input name="email" type="text">
+  <input type="submit" value="Add Student">
+</form>
+```
+
+When a user enters text in the form controls the DOM stores the values of the form fields.  However we can do a lot more combining it with React.
+
+First we will build a React component to manage creating new students.
+
+```javascript
+import React, { Component } from 'react';
+import  './new_student_form.css';
+
+class NewStudentForm extends Component {
+  constructor(props) {
+    super(props);
+}
+render() {
+    return (
+      <div>
+        <form className="new-student-form">
+          <div>
+            <label htmlFor="name">Name:</label>
+            <input
+              name="name" />
+          </div>
+          <div>
+            <label htmlFor="email">Email:</label>
+            <input name="email" />
+          </div>
+          <input className="button success" type="submit" value="Add Student" />
+        </form>
+      </div>
+    );
+  }
+}
+
+export default NewStudentForm;
+```
+
+Now we have a React component to render the form.  It functions just like the HTML form, but can be imported into other React containers.  We term the name and email input fields _Uncontrolled components_ because the DOM is managing the values in the fields.  However now that we have them in a React component we can modify the form controls to have React monitor the input fields and handling the form submission.  To do so we will turn the inputs for name and email into _controlled components_.
```

Comment:
imported into React containers or components?

Reply:
After going through the whole set of notes and then returning back to this discussion of controlled vs uncontrolled components, i'm not sure that we should present it right at the beginning. Could we demonstrate the pattern and then at the end give a name to it instead?